File: vbnet/allergies/Allergies.vb
Imports System.Linq
Imports System.Collections.Generic

Public Class Allergies
    Private ReadOnly _allergies As List(Of String)
    Private ReadOnly _allergens As List(Of String) = New List(Of String)({"eggs", "peanuts", "shellfish", "strawberries", "tomatoes", "chocolate", "pollen", "cats"})

    Public Sub New(codedAllergies As Integer)
        _allergies = _allergens.
            Where(Function(allergen, shiftLeft) (codedAllergies And 1 << shiftLeft) <> 0).
            Select(Function(allergen) allergen.ToString().ToLower()).
            ToList()
    End Sub

    Public Function List() As List(Of String)
        return _allergies
    End Function

    Public Function AllergicTo(allergen As String) As Boolean
        return _allergies.Contains(allergen)
    End Function
End Class

File: prolog/allergies/allergies.pl
allergic_to(Score, Allergen) :- 
    score(Allergen, AllergenScore),
    Score /\ AllergenScore =:= AllergenScore.

allergies(Score, Allergies) :-
    findall(Allergen, allergic_to(Score, Allergen), Allergies).

score(eggs, 1).
score(peanuts, 2).
score(shellfish, 4).
score(strawberries, 8).
score(tomatoes, 16).
score(chocolate, 32).
score(pollen, 64).
score(cats, 128).
